src/manager/node.c

```diff
@@ -799,31 +884,33 @@ static int node_disconnected(UNUSED sd_bus_message *message, void *userdata, UNU
                 }
 
                 usubs->loaded = false;
+
                 UnitSubscription *usub = NULL;
                 UnitSubscription *next_usub = NULL;
+                LIST_FOREACH_SAFE(subs, usub, next_usub, usubs->subs) {
+                        Subscription *sub = usub->sub;
 
-                if (send_state_change) {
-                        LIST_FOREACH_SAFE(subs, usub, next_usub, usubs->subs) {
-                                Subscription *sub = usub->sub;
-                                int r = sub->handle_unit_state_changed(
-                                                sub->monitor,
-                                                node->name,
-                                                sub->unit,
-                                                active_state_to_string(usubs->active_state),
-                                                usubs->substate,
-                                                "virtual");
+                        SubscribedUnit *unit_sub = NULL;
+                        SubscribedUnit *next_unit_sub = NULL;
+                        LIST_FOREACH_SAFE(units, unit_sub, next_unit_sub, sub->subscribed_units) {
+                                int r = sub->handle_unit_removed(
+                                                sub->monitor, node->name, unit_sub->name, "virtual");
                                 if (r < 0) {
                                         hirte_log_error("Failed to emit UnitRemoved signal");
                                 }
-                        }
-                }
 
-                LIST_FOREACH_SAFE(subs, usub, next_usub, usubs->subs) {
-                        Subscription *sub = usub->sub;
-
-                        int r = sub->handle_unit_removed(sub->monitor, node->name, sub->unit, "virtual");
-                        if (r < 0) {
-                                hirte_log_error("Failed to emit UnitRemoved signal");
+                                if (send_state_change) {
```

Comment:
This changes the order of the unit removed and the state changed events.
Also, I think i had some reentrency issue in  the case where handle_unit_removed destroyed the monitor, which made me do the double LIST_FOREACH_SAFE loops that you merged here.

Reply:
Ups, you are right - I didn't think about the order here. Reverted it.